android/settings.gradle

```diff
@@ -57,6 +58,7 @@ project(":expo-random").projectDir = new File("../packages/expo-random/android")
 useExpoModules([
     exclude : [
         'expo-module-template',
+        'react-native-reanimated',
```

Comment:
Did the auto-linking find the `reaniamted` from node_modules? I'm curious why we need to exclude it.

Reply:
`react-native-reanimated` is linkable by `expo-modules-autolinking` since https://github.com/software-mansion/react-native-reanimated/pull/2595. Its vendored code is part of Expo Go codebase, so we need to make sure we don't additionally link the package from hoisted node_modules.

Reply:
exactly! thanks @tsapeta for explaining 💛

Reply:
@Kudo @tsapeta 
Do I need to exclude react-native-reanimated from my bare project? 
When I upgraded react-native-reanimated from 1.x to 2.4.1, I encountered some problems

```
2022-03-11 16:06:53.274 713-851/? E/unknown:NativeModuleInitError: Failed to create NativeModule "UIManager"
    java.lang.NullPointerException
        at java.util.Objects.requireNonNull(Objects.java:220)
        at expo.modules.adapters.react.ModuleRegistryAdapter.createViewManagers(ModuleRegistryAdapter.java:98)
        at expo.modules.ExpoModulesPackage.createViewManagers(ExpoModulesPackage.kt:39)
        at com.facebook.react.ReactInstanceManager.getOrCreateViewManagers(ReactInstanceManager.java:928)
        at com.swmansion.reanimated.ReanimatedPackage.createUIManager(ReanimatedPackage.java:77)
        at com.swmansion.reanimated.ReanimatedPackage.getModule(ReanimatedPackage.java:30)
        at com.facebook.react.TurboReactPackage$ModuleHolderProvider.get(TurboReactPackage.java:159)
        at com.facebook.react.TurboReactPackage$ModuleHolderProvider.get(TurboReactPackage.java:147)
        at com.facebook.react.bridge.ModuleHolder.create(ModuleHolder.java:191)
        at com.facebook.react.bridge.ModuleHolder.getModule(ModuleHolder.java:156)
        at com.facebook.react.bridge.NativeModuleRegistry.getModule(NativeModuleRegistry.java:175)
        at com.facebook.react.bridge.CatalystInstanceImpl.getNativeModule(CatalystInstanceImpl.java:502)
        at com.facebook.react.bridge.CatalystInstanceImpl.getNativeModule(CatalystInstanceImpl.java:478)
        at com.facebook.react.bridge.ReactContext.getNativeModule(ReactContext.java:177)
        at com.swmansion.reanimated.NodesManager.<init>(NodesManager.java:159)
        at com.swmansion.reanimated.ReanimatedModule.getNodesManager(ReanimatedModule.java:99)
        at com.swmansion.reanimated.EXReanimatedAdapter.registerJSIModules(EXReanimatedAdapter.java:9)
        at expo.modules.adapters.reanimated.EXReanimatedPackage$1.onRegisterJSIModules(EXReanimatedPackage.java:21)
        at expo.modules.ReactNativeHostWrapper$JSIModuleContainerPackage.getJSIModules(ReactNativeHostWrapper.kt:95)
        at com.facebook.react.ReactInstanceManager.createReactContext(ReactInstanceManager.java:1376)
        at com.facebook.react.ReactInstanceManager.access$1200(ReactInstanceManager.java:138)
        at com.facebook.react.ReactInstanceManager$5.run(ReactInstanceManager.java:1104)
        at java.lang.Thread.run(Thread.java:923)
2022-03-11 16:06:53.274 713-851/? E/unknown:ReactNative: Exception in native call
    java.lang.NullPointerException
        at java.util.Objects.requireNonNull(Objects.java:220)
        at expo.modules.adapters.react.ModuleRegistryAdapter.createViewManagers(ModuleRegistryAdapter.java:98)
        at expo.modules.ExpoModulesPackage.createViewManagers(ExpoModulesPackage.kt:39)
        at com.facebook.react.ReactInstanceManager.getOrCreateViewManagers(ReactInstanceManager.java:928)
        at com.swmansion.reanimated.ReanimatedPackage.createUIManager(ReanimatedPackage.java:77)
        at com.swmansion.reanimated.ReanimatedPackage.getModule(ReanimatedPackage.java:30)
        at com.facebook.react.TurboReactPackage$ModuleHolderProvider.get(TurboReactPackage.java:159)
        at com.facebook.react.TurboReactPackage$ModuleHolderProvider.get(TurboReactPackage.java:147)
        at com.facebook.react.bridge.ModuleHolder.create(ModuleHolder.java:191)
        at com.facebook.react.bridge.ModuleHolder.getModule(ModuleHolder.java:156)
        at com.facebook.react.bridge.NativeModuleRegistry.getModule(NativeModuleRegistry.java:175)
        at com.facebook.react.bridge.CatalystInstanceImpl.getNativeModule(CatalystInstanceImpl.java:502)
        at com.facebook.react.bridge.CatalystInstanceImpl.getNativeModule(CatalystInstanceImpl.java:478)
        at com.facebook.react.bridge.ReactContext.getNativeModule(ReactContext.java:177)
        at com.swmansion.reanimated.NodesManager.<init>(NodesManager.java:159)
        at com.swmansion.reanimated.ReanimatedModule.getNodesManager(ReanimatedModule.java:99)
        at com.swmansion.reanimated.EXReanimatedAdapter.registerJSIModules(EXReanimatedAdapter.java:9)
        at expo.modules.adapters.reanimated.EXReanimatedPackage$1.onRegisterJSIModules(EXReanimatedPackage.java:21)
        at expo.modules.ReactNativeHostWrapper$JSIModuleContainerPackage.getJSIModules(ReactNativeHostWrapper.kt:95)
        at com.facebook.react.ReactInstanceManager.createReactContext(ReactInstanceManager.java:1376)
        at com.facebook.react.ReactInstanceManager.access$1200(ReactInstanceManager.java:138)
        at com.facebook.react.ReactInstanceManager$5.run(ReactInstanceManager.java:1104)
        at java.lang.Thread.run(Thread.java:923)
```

Reply:
@zhiqingchen could you share you MainApplication / MainActivity as well as expo sdk version?

Reply:
@Kudo yes.

https://github.com/wuba/taro-playground/blob/feat/rn67-upgrade-expo/android/app/src/main/java/com/tarodemo/MainApplication.java

This branch is being upgraded expo, the code is for reference only, it still has some errors.

ps. version react-native-reanimated 2.2.4 is working fine.